module-03-react/src/components/Gallery.jsx

```diff
@@ -0,0 +1,146 @@
+import React, {
+    useEffect,
+    useState,
+    useImperativeHandle,
+    forwardRef,
+} from "react";
+import axios from "axios";
+import "/src/styles/components.css";
+import Card from "./Card.jsx";
+
+const baseURL = "https://swapi.dev/api/";
+
+const Gallery = forwardRef(
+    ({ setImg, search, setSearch, url, setOpenModal, setChoosedCard }, ref) => {
+        const [cards, setCards] = useState([]);
+        const [allResults, setAllResults] = useState([]);
+        const [page, setPage] = useState(1);
+
+        const [loading, setLoading] = useState(false);
+        const [error, setError] = useState(null);
+
+        const bottomButton = document.querySelector(".bottom__btn");
+        let cardCounter = 0;
+
+        useEffect(() => {
+            const bottomButton = document.querySelector(".bottom__btn");
+
+            const getAPI = async () => {
+                setLoading(true);
+                setError(null);
+
+                let allResults = [];
+                let nextPage = baseURL + url;
+                try {
+                    while (nextPage) {
+                        try {
+                            const response = await axios.get(nextPage);
+                            const { results, next } = response.data;
+                            allResults = [...allResults, ...results];
+                            nextPage = next;
+                        } catch (error) {
+                            nextPage = null;
+                        }
+                    }
+                    setAllResults(allResults);
+                    setCards(allResults.slice(0, 15));
+                    if (allResults.length <= 15 && bottomButton) {
+                        bottomButton.classList.add("is-hidden");
+                    } else {
+                        bottomButton.classList.remove("is-hidden");
+                    }
+                    setSearch("");
+                } catch (error) {
+                    setError("Error: can't get data.");
+                } finally {
+                    setLoading(false);
+                }
+            };
+            getAPI();
+            setPage(1);
+        }, [url]);
+
+        useImperativeHandle(ref, () => ({
+            loadMore() {
+                cardCounter = page + 1;
+                setCards(allResults.slice(0, cardCounter * 15));
+                setPage(cardCounter);
+                if (cardCounter * 15 >= allResults.length) {
+                    bottomButton.classList.add("is-hidden");
+                } else {
+                    bottomButton.classList.remove("is-hidden");
+                }
+            },
+        }));
+        return (
+            <section className="gallery">
+                <div className="gallery__grid">
+                    {loading && <div className="loading">Loading...</div>}
+                    {!loading && !error &&
+                        allResults
```

Comment:
The filtering logic inside the render function recalculates on every render, which is inefficient. Consider using `useMemo` to memoize the filtered results and prevent unnecessary recalculations:

Then use memo result directly in your JSX. This optimization will significantly improve performance, especially with larger datasets, by only recalculating when the dependencies change.